tagsets/orsee_mysql.php

```diff
@@ -247,5 +247,39 @@ function dump_array($array,$title="",$dolang=true) {
     echo '</TABLE>';
 }
 
+function version_to_integer($version) {
+    $varray=explode(".",$version);
+    $version_int=$varray[0]*10000+$varray[1]*100+$varray[2]*1;
+    return $version_int;
+}
+
+function check_database_upgrade() {
+    global $settings, $system__version;
+
+    if (!isset($settings['database_version'])) {
+        $settings['database_version']='3.0.2';
+    }
+    $current_system_version_int=version_to_integer($system__version);
+    $db_version_int=version_to_integer($settings['database_version']);
+    if ($db_version_int<$current_system_version_int) {
+        $done=or_upgrade_database($old_db_version_int);
+        $done=orsee_db_save_array(array('option_value'=>$system__version),'options',1,'option_id');
+        return $done;
+    } else {
+        return false;
+    }
+}
+
+function or_upgrade_database($old_version_int) {
+    global $settings;
+    if ($old_version<30003) {
```

Comment:
This area may leave DB in the inconsistent state if any of listed upgrade sections will fail for some reason. I would suggest to use transaction here, so that if anything fails, DB would look is if we did not touch it. Using transaction would require to check that selected MySQL engine supports it (by default is it supported these days, but in ORSEE case there might be old sites with myISAM), perhaps we need a new set of functions or_database_transactionstart, or_database_transactioncommit, or_database_transactionrollback). Alternative option (or the one we can use now as it is quick) would be to update 'database_version' at the end of each if block.

Reply:
I agree that we should put a little more thought in it, such that this allows for incremental database upgrades as well (not only per new version), tracks if there are errors in any upgrade portion, etc. So I will sleep on this a little and not yet include this (and the other feature based on this) in the upcoming minor new release.

Re transactions: ORSEE currently uses myISAM tables, and would need to switch to InnoDB for transaction functionality. This would be a big step and needs to be carefully considered. That said, there is already a transaction function in tagsets/orsee_mysql.php called pdo_transaction($queries).